### PR TITLE
[WIP] Add zeitwerk code loader

### PR DESCRIPTION
**Pull request by jrafanie**

### CODE REVIEW

File: app/controllers/api/base_controller.rb [hunk 1]
```diff
@@ -7,16 +12,16 @@ class BaseController < ActionController::API
...
+    include Parameters
+    include Parser
+    include Manager
+    include Action
+    include Logger
+    include Normalizer
+    include Renderer
+    include Results
+    include Generic
+    include Authentication
```
Comment (Fryguy) [inline app/controllers/api/base_controller.rb:24]: 😍  This is what we've been waiting for!

File: lib/manageiq-api.rb [hunk 2]
```diff
@@ -1,2 +1,25 @@
-require "manageiq/api/engine"
-require "manageiq/api/version"
+
+require "zeitwerk"
+loader = Zeitwerk::Loader.for_gem(warn_on_extra_files: false)
+loader.ignore(__FILE__)
+
+# Add non-standard root directory to the path
+loader.push_dir("#{__dir__}/../app/controllers")
```
Comment (Fryguy) [inline lib/manageiq-api.rb:7]: I would expect any rails engine to have this (as well as things like models, etc - do we have to do this in every plugin?

File: lib/manageiq-api.rb [hunk 3]
```diff
@@ -1,2 +1,25 @@
...
+
+# This tells the loader to not expect Services namespace
+loader.collapse("#{__dir__}/services")
+
+loader.ignore("#{__dir__}/api/api_config.rb")  # Dynamically loads all Vmdb::Plugins - assumes we're mounted in manageiq
+
+# These inflectors teach the loader our naming convention here
+loader.inflector.inflect(
+  "manageiq"  => "ManageIQ",
```
Comment (Fryguy) [inline lib/manageiq-api.rb:16]: Similarly, this is a pain to do everywhere - wonder if there's a way to inherit this from core (perhaps when core is finally converted we can come back and remove the redundant stuff)

File: manageiq-api.gemspec [hunk 4]
```diff
@@ -22,6 +22,10 @@ Gem::Specification.new do |spec|
   # 2.9.0 regressed serialization
   # See: https://github.com/rails/jbuilder/issues/461
   spec.add_dependency "jbuilder", "~> 2.5", "!= 2.9.0"
+  spec.add_dependency "multi_json"
+  spec.add_dependency "rails"
+  spec.add_dependency "responders"
+  spec.add_dependency "zeitwerk"
```
Comment (Fryguy) [inline manageiq-api.gemspec:28]: Also here - not loving all the extra requirements (I realize this is WIP, so not really complaining yet)

File: lib/manageiq-api.rb (hunk 2 above)
Comment (jrafanie) [reply lib/manageiq-api.rb:7]: I'll look.  Thanks.

File: lib/manageiq-api.rb (hunk 3 above)
Comment (jrafanie) [reply lib/manageiq-api.rb:16]: yeah, that's why I wanted to try out a few repos to see what's common or different and to see if there's a better way to do it.

File: manageiq-api.gemspec (hunk 4 above)
Comment (jrafanie) [reply manageiq-api.gemspec:28]: yeah, some were missing dependencies that are only obvious when I tested zeitwerk by enabling eager loading for nearly everything that was fairly easy to autoload.

I can look at extracting the missing dependencies and get them merged on master first.

**Comment by jrafanie** [comment]
I'm closing this one.  Rails engines shouldn't be loading with zeitwerk independently because they get plugged into applications which are responsible for loading code.

If we want to eager load or do this in standalone gems, we can do some of these changes as we get closer to using zeitwerk.

**Pull request closed without merging**